openedx/core/djangoapps/external_user_ids/models.py

```diff
@@ -103,3 +103,38 @@ def add_new_user_id(cls, user, type_name):
                 )
             )
         return external_id, created
+
+    @classmethod
+    def batch_get_or_create(cls, users, type_name):
+        """
+        Create ExternalIds in batch.
+        Arguments:
+            users: List of User to create the IDs for
+            type_name (str): Name of the type of ExternalId
+        Returns:
+            list(ExternalId): Returns the list of external id that was created or retrieved
+        """
+        try:
+            type_obj = ExternalIdType.objects.get(name=type_name)
+        except ExternalIdType.DoesNotExist:
+            LOGGER.info(
+                'Batch ID Creation failed, no external id type of {type}'.format(
+                    type=type_name
+                )
+            )
+            return None
+
+        # get external ids that already exists
+        result = list(cls.objects.filter(user__in=users, external_id_type=type_obj))
+
+        # find users for those external ids needs to be created
+        users_wo_externalid = set(users) - {eid.user for eid in result}
```

Comment:
I think this is not the optimal solution, given that a potentially big user dataset might come in.
Forcing a queryset evaluation and then subtracting two user objects sets might be an expensive operation.

Reply:
@giovannicimolin, I've optimized it to use a query instead of set operation.